System.OverflowException for ParseInt32[Bug]

**Describe the bug**
When gitversion tries to parse a prerelease tag followed by a 10+ digit number an exception is thrown.

## Expected Behavior
Gitversion should catch the exception and handle it some other way

## Actual Behavior
Gitversion doesn't handle any exceptions inside a method called TryParse

## Possible Fix
Make TryParse method of `GitVersionCore/VersionCalculation/SemanticVersioning/SemanticVersion.cs` actually use a try-catch as the name would suggest.
Alternatively have `GitVersionCore/VersionCalculation/SemanticVersioning/SemanticVersionPreReleaseTag.cs` Parse method use a TryParse for parsing the Int

## Steps to Reproduce
Use a tag with a long number in it. Ours has been automated.
Here's the exception along with the tag that likely caused the issue:
```
System.OverflowException: Value was either too large or too small for an Int32.
   at System.Number.ParseInt32(String s, NumberStyles style, NumberFormatInfo info)
   at GitVersion.SemanticVersionPreReleaseTag.Parse(String preReleaseTag)
   at GitVersion.SemanticVersion.TryParse(String version, String tagPrefixRegex, SemanticVersion& semanticVersion)
   at GitVersion.GitRepoMetadataProvider.GetValidVersionTags(IRepository repository, String tagPrefixRegex, Nullable`1 olderThan)
   at GitVersion.VersionCalculation.BaseVersionCalculators.TaggedCommitVersionStrategy.GetTaggedVersions(GitVersionContext context, Branch currentBranch, Nullable`1 olderThan)
   at GitVersion.VersionCalculation.BaseVersionCalculators.TaggedCommitVersionStrategy.GetVersions(GitVersionContext context)
   at System.Linq.Enumerable.<SelectManyIterator>d__17`2.MoveNext()
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at GitVersion.VersionCalculation.BaseVersionCalculator.GetBaseVersion(GitVersionContext context)
   at GitVersion.VersionCalculation.NextVersionCalculator.FindVersion(GitVersionContext context)
   at GitVersion.GitVersionFinder.FindVersion(GitVersionContext context)
   at GitVersion.ExecuteCore.<>c__DisplayClass7_0.<ExecuteInternal>b__0(IRepository repo)
   at GitVersion.GitPreparer.WithRepository[TResult](Func`2 action)
   at GitVersion.ExecuteCore.ExecuteGitVersion(String targetUrl, String dynamicRepositoryLocation, Authentication authentication, String targetBranch, Boolean noFetch, String workingDirectory, String commitId, Config overrideConfig, Boolean noCache)
   at GitVersion.SpecifiedArgumentRunner.Run(Arguments arguments, IFileSystem fileSystem)
   at GitVersion.Program.VerifyArgumentsAndRun()
  INFO [08/21/20 7:08:34:93] 
  INFO [08/21/20 7:08:34:93] Attempting to show the current git graph (please include in issue): 
  INFO [08/21/20 7:08:34:93] Showing max of 100 commits
  INFO [08/21/20 7:08:34:99] *   a401875 18 hours ago  (HEAD -> pull/22012/merge, pull/22012/merge)
|\  
| * 38895d1 18 hours ago  (origin/573637-fix-gitversion-issue, 573637-fix-gitversion-issue)
| * be84a2e 20 hours ago 
|/  
| *   5a4a29d 21 hours ago  (tag: 5.2.14, origin/release, release)
| |\  
| |/  
|/|   
| | *   513ba3f 4 days ago  (tag: 5.2.14-PullRequest2189021890, tag: 5.2.14-PullRequest21890.5)
| | |\ 
```

## Context
Our CI build pipeline fails and we are unable to rebuild

## Your Environment
Azure DevOps Task. Gitversion is 5.01 there, however this issue is also present in more recent versions.

just hit similar error and getting it consistently on azure devops, but can't replicate locally, task is using 5.6.4, but was able to replicate it with 5.6.3 as well
```
      - task: gitversion/setup@0
        inputs:
          versionSpec: '5.6.4'

      - task: gitversion/execute@0
        displayName: GitVersion Diag
        inputs:
          useConfigFile: true
          configFilePath: .\Templates\ComponentVersioning\GitVersion.yml
          additionalArguments: /targetpath $(Build.SourcesDirectory)\Source /diag
```
errors:
```
| *   7e99417 2 years, 7 months ago  (tag: V0.1.0, tag: 1.0.0)
| |\  
| |/  
|/|   
  ERROR [02/03/21 9:33:11:56] An unexpected error occurred:
System.OverflowException: Value was either too large or too small for an Int32.
   at System.Number.ThrowOverflowOrFormatException(ParsingStatus status, TypeCode type)
   at System.Number.ParseInt32(ReadOnlySpan`1 value, NumberStyles styles, NumberFormatInfo info)
   at System.Int32.Parse(String s)
   at GitVersion.SemanticVersionPreReleaseTag.Parse(String preReleaseTag) in D:\a\GitVersion\GitVersion\src\GitVersionCore\VersionCalculation\SemanticVersioning\SemanticVersionPreReleaseTag.cs:line 106
   at GitVersion.SemanticVersion.TryParse(String version, String tagPrefixRegex, SemanticVersion& semanticVersion) in D:\a\GitVersion\GitVersion\src\GitVersionCore\VersionCalculation\SemanticVersioning\SemanticVersion.cs:line 173
   at GitVersion.RepositoryMetadataProvider.<>c__DisplayClass23_0.<GetCurrentCommitTaggedVersion>b__0(ITag t) in D:\a\GitVersion\GitVersion\src\GitVersionCore\Core\RepositoryMetadataProvider.cs:line 322
   at System.Linq.Enumerable.SelectManySingleSelectorIterator`2.MoveNext()
   at System.Linq.Enumerable.Max[TSource](IEnumerable`1 source)
   at GitVersion.RepositoryMetadataProvider.GetCurrentCommitTaggedVersion(ICommit commit, EffectiveConfiguration config) in D:\a\GitVersion\GitVersion\src\GitVersionCore\Core\RepositoryMetadataProvider.cs:line 318
   at GitVersion.GitVersionContextFactory.Create(GitVersionOptions gitVersionOptions) in D:\a\GitVersion\GitVersion\src\GitVersionCore\Core\GitVersionContextFactory.cs:line 42
   at GitVersion.GitVersionCoreModule.<>c__DisplayClass0_0.<RegisterTypes>b__3() in D:\a\GitVersion\GitVersion\src\GitVersionCore\GitVersionCoreModule.cs:line 61
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.get_Value()
   at GitVersion.VersionCalculation.NextVersionCalculator.get_context() in D:\a\GitVersion\GitVersion\src\GitVersionCore\VersionCalculation\NextVersionCalculator.cs:line 17
   at GitVersion.VersionCalculation.NextVersionCalculator.FindVersion() in D:\a\GitVersion\GitVersion\src\GitVersionCore\VersionCalculation\NextVersionCalculator.cs:line 32
   at GitVersion.GitVersionCalculateTool.CalculateVersionVariables() in D:\a\GitVersion\GitVersion\src\GitVersionCore\Core\GitVersionCalculateTool.cs:line 52
   at GitVersion.GitVersionExecutor.RunGitVersionTool(GitVersionOptions gitVersionOptions) in D:\a\GitVersion\GitVersion\src\GitVersionExe\GitVersionExecutor.cs:line 61
* |   22258d9 2 years, 7 months ago 
|\ \  
| * | 6a5fe7e 2 years, 7 months ago 
```
later on
```
INFO [02/03/21 9:33:11:17] End: Attempting to inherit branch configuration from parent branch (Took: 57.26ms)
  ERROR [02/03/21 9:33:11:56] An unexpected error occurred:
System.OverflowException: Value was either too large or too small for an Int32.
   at System.Number.ThrowOverflowOrFormatException(ParsingStatus status, TypeCode type)
   at System.Number.ParseInt32(ReadOnlySpan`1 value, NumberStyles styles, NumberFormatInfo info)
   at System.Int32.Parse(String s)
   at GitVersion.SemanticVersionPreReleaseTag.Parse(String preReleaseTag) in D:\a\GitVersion\GitVersion\src\GitVersionCore\VersionCalculation\SemanticVersioning\SemanticVersionPreReleaseTag.cs:line 106
   at GitVersion.SemanticVersion.TryParse(String version, String tagPrefixRegex, SemanticVersion& semanticVersion) in D:\a\GitVersion\GitVersion\src\GitVersionCore\VersionCalculation\SemanticVersioning\SemanticVersion.cs:line 173
   at GitVersion.RepositoryMetadataProvider.<>c__DisplayClass23_0.<GetCurrentCommitTaggedVersion>b__0(ITag t) in D:\a\GitVersion\GitVersion\src\GitVersionCore\Core\RepositoryMetadataProvider.cs:line 322
   at System.Linq.Enumerable.SelectManySingleSelectorIterator`2.MoveNext()
   at System.Linq.Enumerable.Max[TSource](IEnumerable`1 source)
   at GitVersion.RepositoryMetadataProvider.GetCurrentCommitTaggedVersion(ICommit commit, EffectiveConfiguration config) in D:\a\GitVersion\GitVersion\src\GitVersionCore\Core\RepositoryMetadataProvider.cs:line 318
   at GitVersion.GitVersionContextFactory.Create(GitVersionOptions gitVersionOptions) in D:\a\GitVersion\GitVersion\src\GitVersionCore\Core\GitVersionContextFactory.cs:line 42
   at GitVersion.GitVersionCoreModule.<>c__DisplayClass0_0.<RegisterTypes>b__3() in D:\a\GitVersion\GitVersion\src\GitVersionCore\GitVersionCoreModule.cs:line 61
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.get_Value()
   at GitVersion.VersionCalculation.NextVersionCalculator.get_context() in D:\a\GitVersion\GitVersion\src\GitVersionCore\VersionCalculation\NextVersionCalculator.cs:line 17
   at GitVersion.VersionCalculation.NextVersionCalculator.FindVersion() in D:\a\GitVersion\GitVersion\src\GitVersionCore\VersionCalculation\NextVersionCalculator.cs:line 32
   at GitVersion.GitVersionCalculateTool.CalculateVersionVariables() in D:\a\GitVersion\GitVersion\src\GitVersionCore\Core\GitVersionCalculateTool.cs:line 52
   at GitVersion.GitVersionExecutor.RunGitVersionTool(GitVersionOptions gitVersionOptions) in D:\a\GitVersion\GitVersion\src\GitVersionExe\GitVersionExecutor.cs:line 61
  INFO [02/03/21 9:33:11:56] Attempting to show the current git graph (please include in issue): 
  INFO [02/03/21 9:33:11:56] Showing max of 100 commits
```
and at the end
```
  INFO [02/03/21 9:33:11:62] Done writing 
##[debug]task result: Failed
##[error]SyntaxError: Unexpected end of JSON input
##[debug]Processed: ##vso[task.issue type=error;]SyntaxError: Unexpected end of JSON input
##[debug]Processed: ##vso[task.complete result=Failed;done=true;]SyntaxError: Unexpected end of JSON input
##[section]Finishing: GitVersion Diag
```

--- a/src/GitVersion.Core.Tests/VersionCalculation/SemanticVersionTests.cs
+++ b/src/GitVersion.Core.Tests/VersionCalculation/SemanticVersionTests.cs
@@ -35,8 +35,10 @@ public class SemanticVersionTests : TestBase
     [TestCase("version-1.2.3", 1, 2, 3, null, null, null, null, null, null, "1.2.3", "version-")]
     [TestCase("1", 1, 0, 0, null, null, null, null, null, null, "1.0.0", null)]
     [TestCase("1.1", 1, 1, 0, null, null, null, null, null, null, "1.1.0", null)]
+    [TestCase("1.0.0-develop-20201007113711", 1, 0, 0, "develop-20201007113711", null, null, null, null, null, "1.0.0-develop-20201007113711", null)]
+    [TestCase("20201007113711.658165168461351.64136516984163213-develop-20201007113711.98848747823+65416321321", 20201007113711, 658165168461351, 64136516984163213, "develop-20201007113711", 98848747823, 65416321321, null, null, null, "20201007113711.658165168461351.64136516984163213-develop-20201007113711.98848747823+65416321321", null)]
     public void ValidateVersionParsing(
-        string versionString, int major, int minor, int patch, string tag, int? tagNumber, int? numberOfBuilds,
+        string versionString, long major, long minor, long patch, string tag, long? tagNumber, long? numberOfBuilds,
         string branchName, string sha, string otherMetaData, string fullFormattedVersionString, string tagPrefixRegex)
     {
         fullFormattedVersionString ??= versionString;

--- a/src/GitVersion.Core/VersionCalculation/NextVersionCalculator.cs
+++ b/src/GitVersion.Core/VersionCalculation/NextVersionCalculator.cs
@@ -117,7 +117,7 @@ public class NextVersionCalculator : INextVersionCalculator
     {
         var tagToUse = context.Configuration?.GetBranchSpecificTag(this.log, context.CurrentBranch?.Name.Friendly, branchNameOverride);
 
-        int? number = null;
+        long? number = null;
 
         var lastTag = this.repositoryStore
             .GetVersionTagsOnBranch(context.CurrentBranch!, context.Configuration?.GitTagPrefix)

--- a/src/GitVersion.Core/VersionCalculation/SemanticVersioning/SemanticVersion.cs
+++ b/src/GitVersion.Core/VersionCalculation/SemanticVersioning/SemanticVersion.cs
@@ -13,13 +13,13 @@ public class SemanticVersion : IFormattable, IComparable<SemanticVersion>, IEqua
         @"^(?<SemVer>(?<Major>\d+)(\.(?<Minor>\d+))?(\.(?<Patch>\d+))?)(\.(?<FourthPart>\d+))?(-(?<Tag>[^\+]*))?(\+(?<BuildMetaData>.*))?$",
         RegexOptions.Compiled);
 
-    public int Major;
-    public int Minor;
-    public int Patch;
+    public long Major;
+    public long Minor;
+    public long Patch;
     public SemanticVersionPreReleaseTag? PreReleaseTag;
     public SemanticVersionBuildMetaData? BuildMetaData;
 
-    public SemanticVersion(int major = 0, int minor = 0, int patch = 0)
+    public SemanticVersion(long major = 0, long minor = 0, long patch = 0)
     {
         this.Major = major;
         this.Minor = minor;
@@ -74,9 +74,9 @@ public class SemanticVersion : IFormattable, IComparable<SemanticVersion>, IEqua
     {
         unchecked
         {
-            var hashCode = this.Major;
-            hashCode = (hashCode * 397) ^ this.Minor;
-            hashCode = (hashCode * 397) ^ this.Patch;
+            var hashCode = this.Major.GetHashCode();
+            hashCode = (hashCode * 397) ^ this.Minor.GetHashCode();
+            hashCode = (hashCode * 397) ^ this.Patch.GetHashCode();
             hashCode = (hashCode * 397) ^ (this.PreReleaseTag != null ? this.PreReleaseTag.GetHashCode() : 0);
             hashCode = (hashCode * 397) ^ (this.BuildMetaData != null ? this.BuildMetaData.GetHashCode() : 0);
             return hashCode;
@@ -168,9 +168,9 @@ public class SemanticVersion : IFormattable, IComparable<SemanticVersion>, IEqua
 
         semanticVersion = new SemanticVersion
         {
-            Major = int.Parse(parsed.Groups["Major"].Value),
-            Minor = parsed.Groups["Minor"].Success ? int.Parse(parsed.Groups["Minor"].Value) : 0,
-            Patch = parsed.Groups["Patch"].Success ? int.Parse(parsed.Groups["Patch"].Value) : 0,
+            Major = long.Parse(parsed.Groups["Major"].Value),
+            Minor = parsed.Groups["Minor"].Success ? long.Parse(parsed.Groups["Minor"].Value) : 0,
+            Patch = parsed.Groups["Patch"].Success ? long.Parse(parsed.Groups["Patch"].Value) : 0,
             PreReleaseTag = SemanticVersionPreReleaseTag.Parse(parsed.Groups["Tag"].Value),
             BuildMetaData = semanticVersionBuildMetaData
         };

--- a/src/GitVersion.Core/VersionCalculation/SemanticVersioning/SemanticVersionBuildMetaData.cs
+++ b/src/GitVersion.Core/VersionCalculation/SemanticVersioning/SemanticVersionBuildMetaData.cs
@@ -14,15 +14,15 @@ public class SemanticVersionBuildMetaData : IFormattable, IEquatable<SemanticVer
     private static readonly LambdaEqualityHelper<SemanticVersionBuildMetaData> EqualityHelper =
         new(x => x.CommitsSinceTag, x => x.Branch, x => x.Sha);
 
-    public int? CommitsSinceTag;
+    public long? CommitsSinceTag;
     public string? Branch;
     public string? Sha;
     public string? ShortSha;
     public string? OtherMetaData;
     public DateTimeOffset? CommitDate;
     public string? VersionSourceSha;
-    public int? CommitsSinceVersionSource;
-    public int UncommittedChanges;
+    public long? CommitsSinceVersionSource;
+    public long UncommittedChanges;
 
     public SemanticVersionBuildMetaData()
     {
@@ -124,7 +124,7 @@ public class SemanticVersionBuildMetaData : IFormattable, IEquatable<SemanticVer
 
         if (parsed.Groups["BuildNumber"].Success)
         {
-            semanticVersionBuildMetaData.CommitsSinceTag = int.Parse(parsed.Groups["BuildNumber"].Value);
+            semanticVersionBuildMetaData.CommitsSinceTag = long.Parse(parsed.Groups["BuildNumber"].Value);
             semanticVersionBuildMetaData.CommitsSinceVersionSource = semanticVersionBuildMetaData.CommitsSinceTag ?? 0;
         }
 

--- a/src/GitVersion.Core/VersionCalculation/SemanticVersioning/SemanticVersionPreReleaseTag.cs
+++ b/src/GitVersion.Core/VersionCalculation/SemanticVersioning/SemanticVersionPreReleaseTag.cs
@@ -15,7 +15,7 @@ public class SemanticVersionPreReleaseTag :
     {
     }
 
-    public SemanticVersionPreReleaseTag(string? name, int? number)
+    public SemanticVersionPreReleaseTag(string? name, long? number)
     {
         Name = name;
         Number = number;
@@ -29,7 +29,7 @@ public class SemanticVersionPreReleaseTag :
     }
 
     public string? Name { get; set; }
-    public int? Number { get; set; }
+    public long? Number { get; set; }
     public bool? PromotedFromCommits { get; set; }
 
     public override bool Equals(object? obj) => Equals(obj as SemanticVersionPreReleaseTag);
@@ -76,7 +76,7 @@ public class SemanticVersionPreReleaseTag :
         }
 
         var value = match.Groups["name"].Value;
-        var number = match.Groups["number"].Success ? int.Parse(match.Groups["number"].Value) : (int?)null;
+        var number = match.Groups["number"].Success ? long.Parse(match.Groups["number"].Value) : (long?)null;
         if (value.EndsWith("-"))
             return new SemanticVersionPreReleaseTag(preReleaseTag, null);
 

